### ISSUE TITLE
Filtering for Libretro Core

### ISSUE THREAD

**Issue by Papermanzero** [FCare/Kronos#1424]
What's your frontend ? (standalone, libretro, ...) ?
libretro

What's your OS ? (windows, linux, OSX, ...) ?
win10

Is it possible to add bilinear, jinc2, xbrz filtering for kronos. In a way that it is also available for libretro. Swanstation aka duckstation is a good reference

**Comment by barbudreadmon** [comment]
As answered at https://github.com/libretro/yabause/issues/297#issuecomment-1358562667, the filters were deemed useless in the libretro core because RetroArch already has filters and shaders.

**Comment by Papermanzero** [comment]
not for textures. Otherwise it would be useles for swanstation, mupen64, dolphin, pcsx2, ppsspp, as well😂

**Comment by barbudreadmon** [comment]
I might be wrong, but iirc Kronos's internal filters are applied right before rendering, meaning they don't work differently from RA's features. I'm gonna confirm with team.

**Comment by Papermanzero** [comment]
Please don‘t mix it up with screen filters. With screen filters you are absolutely right. Those would be useless. 

But texture filters are applied in the graphics pipeline before they will be rendered on the polygons. Unfortunately the saturn can only use nearest neighbour in VDP1. That’s why you have those blocky textures in some games like Panzer Dragoon or Nights. 

![CF2E349E-163C-4E76-910A-B4B94847B887](https://user-images.githubusercontent.com/10263536/208628257-77dae6c9-6f0b-40ad-9096-4faabadbef61.jpeg)
![8091DD9B-164C-4644-B7F4-EE418E3BDAF7](https://user-images.githubusercontent.com/10263536/208628265-4ea8ccbf-4d98-4bfa-8d62-23cfc6d68cae.jpeg)

**Comment by barbudreadmon** [comment]
I got the confirmation from @FCare that only screen filters are available right now in Kronos.

It seems it would be quite a challenge to implement texture filters due to saturn's architecture, so it's not a priority right now.

**Comment by Papermanzero** [comment]
Yes, you need to extend the video pipeline in VDP1 with filter functions. It would be great because it improves the picture quality a lot.

**Comment by barbudreadmon** [comment]
>you need to extend the video pipeline in VDP1 with filter functions.

Apparently it might be a bit more challenging than that, because the vdp1 works mostly with lookup tables, not textures.

**Comment by Papermanzero** [comment]
uff, then it is complicated. In general you need to look for the calculated texture and apply the filter on that. This is an extension of the hardware, which is not trivial in saturns case.

**Comment by FCare** [comment]
Not only improve tendering pipeline. Most of vdp1 texture are using
palette. The palette itself is managed by vdp2 but geometry is managed by
vdp1. And both rendering process do not happened at the same time.
It would require to save all details of the texture used in a quad when the
vdp1 render it, per pixel since pixels can be overriding some others.
At the end, before the vdp2 rendering phase, we would have to render full
vdp1 texture, even if only one pixel is used, and the do the texture
filtering during the vdp2 rendering pass.
For sure, it will have a big impact in term of performance and it will
require long days of dev.

Right now, it is not on the top priority list since I am the only dev
working on kronos.

Le mar. 20 déc. 2022, 11:06, Papermanzero ***@***.***> a
écrit :
